In [1]:
import pandas as pd

In [ ]:
!pip uninstall keras

In [2]:
filepath = '05-06-2022-12-32-52-grocery_dataset.txt/grocery_dataset.txt'

In [3]:
df = pd.read_csv(filepath, engine='python',
                header=None, sep='delimiter')
df.columns = ['groceries']
df.head()

groceries
0  citrus fruit,semi-finished bread,margarine,rea...
1                       tropical fruit,yogurt,coffee
2                                         whole milk
3        pip fruit,yogurt,cream cheese ,meat spreads
4  other vegetables,whole milk,condensed milk,lon...

In [4]:
import numpy as np
np.sum(df.isna())

groceries    0
dtype: int64

In [5]:
df.describe()

groceries
count          9835
unique         7011
top     canned beer
freq            260

In [6]:
from mlxtend.preprocessing import TransactionEncoder

In [7]:
data_set = df['groceries'].apply(lambda x: x.split(',')).to_list()
data_set[0:5]

[['citrus fruit', 'semi-finished bread', 'margarine', 'ready soups'],
 ['tropical fruit', 'yogurt', 'coffee'],
 ['whole milk'],
 ['pip fruit', 'yogurt', 'cream cheese ', 'meat spreads'],
 ['other vegetables',
  'whole milk',
  'condensed milk',
  'long life bakery product']]

In [8]:
te = TransactionEncoder()
te_arr = te.fit_transform(data_set, sparse=True)
sparse_df = pd.DataFrame.sparse.from_spmatrix(
                te_arr, columns=[str(c) for c in te.columns_])
sparse_df.head()

Instant food products  UHT-milk  abrasive cleaner  artif. sweetener  \
0                      0         0                 0                 0   
1                      0         0                 0                 0   
2                      0         0                 0                 0   
3                      0         0                 0                 0   
4                      0         0                 0                 0   

   baby cosmetics  baby food  bags  baking powder  bathroom cleaner  beef  \
0               0          0     0              0                 0     0   
1               0          0     0              0                 0     0   
2               0          0     0              0                 0     0   
3               0          0     0              0                 0     0   
4               0          0     0              0                 0     0   

   ...  turkey  vinegar  waffles  whipped/sour cream  whisky  white bread  \
0  ...       0        0        0                   0       0            0   
1  ...       0        0        0                   0       0            0   
2  ...       0        0        0                   0       0            0   
3  ...       0        0        0                   0       0            0   
4  ...       0        0        0                   0       0            0   

   white wine  whole milk  yogurt  zwieback  
0           0           0       0         0  
1           0           0       1         0  
2           0           1       0         0  
3           0           0       1         0  
4           0           1       0         0  

[5 rows x 171 columns]

### Áp dụng apriori cho asociation rule mining

In [9]:
from mlxtend.frequent_patterns import apriori

In [10]:
te = TransactionEncoder()
te_ary = te.fit_transform(data_set, sparse=False)
df = pd.DataFrame(te_ary, columns=te.columns_)
df

Instant food products  UHT-milk  abrasive cleaner  artif. sweetener  \
0                     False     False             False             False   
1                     False     False             False             False   
2                     False     False             False             False   
3                     False     False             False             False   
4                     False     False             False             False   
...                     ...       ...               ...               ...   
9830                  False     False             False             False   
9831                  False     False             False             False   
9832                  False     False             False             False   
9833                  False     False             False             False   
9834                  False     False             False             False   

      baby cosmetics  baby food   bags  baking powder  bathroom cleaner  \
0              False      False  False          False             False   
1              False      False  False          False             False   
2              False      False  False          False             False   
3              False      False  False          False             False   
4              False      False  False          False             False   
...              ...        ...    ...            ...               ...   
9830           False      False  False          False             False   
9831           False      False  False          False             False   
9832           False      False  False          False             False   
9833           False      False  False          False             False   
9834           False      False  False          False             False   

       beef  ...  turkey  vinegar  waffles  whipped/sour cream  whisky  \
0     False  ...   False    False    False               False   False   
1     False  ...   False    False    False               False   False   
2     False  ...   False    False    False               False   False   
3     False  ...   False    False    False               False   False   
4     False  ...   False    False    False               False   False   
...     ...  ...     ...      ...      ...                 ...     ...   
9830   True  ...   False    False    False                True   False   
9831  False  ...   False    False    False               False   False   
9832  False  ...   False    False    False               False   False   
9833  False  ...   False    False    False               False   False   
9834  False  ...   False     True    False               False   False   

      white bread  white wine  whole milk  yogurt  zwieback  
0           False       False       False   False     False  
1           False       False       False    True     False  
2           False       False        True   False     False  
3           False       False       False    True     False  
4           False       False        True   False     False  
...           ...         ...         ...     ...       ...  
9830        False       False        True   False     False  
9831        False       False       False   False     False  
9832        False       False       False    True     False  
9833        False       False       False   False     False  
9834        False       False       False   False     False  

[9835 rows x 171 columns]

In [11]:
freq_items = apriori(df, min_support=0.05, 
                     use_colnames=True)
freq_items

support                        itemsets
0   0.052466                          (beef)
1   0.080529                  (bottled beer)
2   0.110524                 (bottled water)
3   0.064870                   (brown bread)
4   0.055414                        (butter)
5   0.077682                   (canned beer)
6   0.082766                  (citrus fruit)
7   0.058058                        (coffee)
8   0.053279                          (curd)
9   0.063447                 (domestic eggs)
10  0.058973                   (frankfurter)
11  0.072293         (fruit/vegetable juice)
12  0.058566                     (margarine)
13  0.052364                       (napkins)
14  0.079817                    (newspapers)
15  0.193493              (other vegetables)
16  0.088968                        (pastry)
17  0.075648                     (pip fruit)
18  0.057651                          (pork)
19  0.183935                    (rolls/buns)
20  0.108998               (root vegetables)
21  0.093950                       (sausage)
22  0.098526                 (shopping bags)
23  0.174377                          (soda)
24  0.104931                (tropical fruit)
25  0.071683            (whipped/sour cream)
26  0.255516                    (whole milk)
27  0.139502                        (yogurt)
28  0.074835  (other vegetables, whole milk)
29  0.056634        (whole milk, rolls/buns)
30  0.056024            (yogurt, whole milk)

+ Do dữ liệu khá nhỏ gần 10000 rows so với unique values (khoảng 7000 unique values) nên độ hỗ trợ của các tập dữ liệu tương đối thấp
+ Do số lượng unique values khá cao với số lượng rows của dữ liệu, mà tần suất xuất hiện của các unique values khá thấp ở mỗi row nên support của các tập dữ liệu tương đối thấp

In [14]:
# Kiểm tra thử support của một tập dữ liệu
k=0
count=0
for i in data_set:
    for j in i:
        if j=='bottled beer':
            count+=1
count/9834

0.08053691275167785

In [15]:
from mlxtend.frequent_patterns import association_rules

In [16]:
association_rules(freq_items, metric='lift',
                 min_threshold=1.01)

antecedents         consequents  antecedent support  \
0  (other vegetables)        (whole milk)            0.193493   
1        (whole milk)  (other vegetables)            0.255516   
2        (whole milk)        (rolls/buns)            0.255516   
3        (rolls/buns)        (whole milk)            0.183935   
4            (yogurt)        (whole milk)            0.139502   
5        (whole milk)            (yogurt)            0.255516   

   consequent support   support  confidence      lift  leverage  conviction  
0            0.255516  0.074835    0.386758  1.513634  0.025394    1.214013  
1            0.193493  0.074835    0.292877  1.513634  0.025394    1.140548  
2            0.183935  0.056634    0.221647  1.205032  0.009636    1.048452  
3            0.255516  0.056634    0.307905  1.205032  0.009636    1.075696  
4            0.255516  0.056024    0.401603  1.571735  0.020379    1.244132  
5            0.139502  0.056024    0.219260  1.571735  0.020379    1.102157

##### comments:
+ Mọi người mua nhiều nhất (whole milk) khi vào grocery
+ Xác suất khi mua whole milk trong khi trước đó đã mua yogurt là cao nhất (conf = 0.4). Khả năng mua whole milk + yogurt cao gấp 1.57 lần chỉ mua riêng lẻ whole milk hoặc yogurt (lift = 1.57)
+ Nhìn chung độ thuyết phục (conviction) của các tập dữ liệu còn khá thấp nên không tin tưởng để đánh giá sự phụ thuộc của các groceries

### So sánh thời gian apriori và FP-tree chạy mô hình

In [17]:
from mlxtend.frequent_patterns import fpgrowth

In [18]:
te = TransactionEncoder()
te_ary = te.fit_transform(data_set, sparse=False)
df = pd.DataFrame(te_ary, columns=te.columns_)
df

Instant food products  UHT-milk  abrasive cleaner  artif. sweetener  \
0                     False     False             False             False   
1                     False     False             False             False   
2                     False     False             False             False   
3                     False     False             False             False   
4                     False     False             False             False   
...                     ...       ...               ...               ...   
9830                  False     False             False             False   
9831                  False     False             False             False   
9832                  False     False             False             False   
9833                  False     False             False             False   
9834                  False     False             False             False   

      baby cosmetics  baby food   bags  baking powder  bathroom cleaner  \
0              False      False  False          False             False   
1              False      False  False          False             False   
2              False      False  False          False             False   
3              False      False  False          False             False   
4              False      False  False          False             False   
...              ...        ...    ...            ...               ...   
9830           False      False  False          False             False   
9831           False      False  False          False             False   
9832           False      False  False          False             False   
9833           False      False  False          False             False   
9834           False      False  False          False             False   

       beef  ...  turkey  vinegar  waffles  whipped/sour cream  whisky  \
0     False  ...   False    False    False               False   False   
1     False  ...   False    False    False               False   False   
2     False  ...   False    False    False               False   False   
3     False  ...   False    False    False               False   False   
4     False  ...   False    False    False               False   False   
...     ...  ...     ...      ...      ...                 ...     ...   
9830   True  ...   False    False    False                True   False   
9831  False  ...   False    False    False               False   False   
9832  False  ...   False    False    False               False   False   
9833  False  ...   False    False    False               False   False   
9834  False  ...   False     True    False               False   False   

      white bread  white wine  whole milk  yogurt  zwieback  
0           False       False       False   False     False  
1           False       False       False    True     False  
2           False       False        True   False     False  
3           False       False       False    True     False  
4           False       False        True   False     False  
...           ...         ...         ...     ...       ...  
9830        False       False        True   False     False  
9831        False       False       False   False     False  
9832        False       False       False    True     False  
9833        False       False       False   False     False  
9834        False       False       False   False     False  

[9835 rows x 171 columns]

In [19]:
%%time
freq_items = fpgrowth(df, use_colnames=True, min_support=0.05)

CPU times: total: 78.1 ms
Wall time: 77.8 ms


In [20]:
%%time
freq_items = apriori(df, use_colnames=True, min_support=0.05)

CPU times: total: 46.9 ms
Wall time: 31.9 ms


+ Trong tường hợp này apriori có tốc độ xử lý nhanh hơn FP tree